src/Cli/dotnet/commands/dotnet-publish/Program.cs

```diff
@@ -43,30 +64,64 @@ public static PublishCommand FromParseResult(ParseResult parseResult, string msb
-            msbuildArgs.AddRange(slnOrProjectArgs ?? Array.Empty<string>());
-
             bool noRestore = parseResult.HasOption(PublishCommandParser.NoRestoreOption)
                           || parseResult.HasOption(PublishCommandParser.NoBuildOption);
 
+
+            var publishProfileProperties = DiscoverPropertiesFromPublishProfile(parseResult);
+            var standardMSbuildProperties = CreatePropertyListForPublishInvocation(parseResult);
             return new PublishCommand(
-                msbuildArgs,
+                // properties defined by the selected publish profile should override any other properties,
+                // so they should be added after the other properties.
+                standardMSbuildProperties.Concat(publishProfileProperties),
                 noRestore,
-                msbuildPath);
+                msbuildPath
+            );
+
+            List<string> CreatePropertyListForPublishInvocation(ParseResult parseResult) {
+                var msbuildArgs = new List<string>()
+                    {
+                        "-target:Publish",
+                        "--property:_IsPublishing=true" // This property will not hold true for MSBuild /t:Publish. VS should also inject this property when publishing in the future.
+                    };
+
+                IEnumerable<string> slnOrProjectArgs = parseResult.GetValueForArgument(PublishCommandParser.SlnOrProjectArgument);
+
+                CommonOptions.ValidateSelfContainedOptions(parseResult.HasOption(PublishCommandParser.SelfContainedOption),
+                    parseResult.HasOption(PublishCommandParser.NoSelfContainedOption));
+
+                msbuildArgs.AddRange(parseResult.OptionValuesToBeForwarded(PublishCommandParser.GetCommand()));
+                ReleasePropertyProjectLocator projectLocator = new ReleasePropertyProjectLocator(Environment.GetEnvironmentVariable(EnvironmentVariableNames.ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS) != null);
+                msbuildArgs.AddRange(projectLocator.GetCustomDefaultConfigurationValueIfSpecified(parseResult, MSBuildPropertyNames.PUBLISH_RELEASE, slnOrProjectArgs, PublishCommandParser.ConfigurationOption) ?? Array.Empty<string>());
+                msbuildArgs.AddRange(slnOrProjectArgs ?? Array.Empty<string>());
+
+                return msbuildArgs;
+            }
+
+            List<string> DiscoverPropertiesFromPublishProfile(ParseResult parseResult)
+            {
+                ReleasePropertyProjectLocator projectLocator = new ReleasePropertyProjectLocator(Environment.GetEnvironmentVariable(EnvironmentVariableNames.ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS) != null);
+                var cliProps = projectLocator.GetGlobalPropertiesFromUserArgs(parseResult);
+                var projectInstance = projectLocator.GetTargetedProject(parseResult.GetValueForArgument(PublishCommandParser.SlnOrProjectArgument), cliProps, MSBuildPropertyNames.PUBLISH_RELEASE);
```

Comment:
My main concern here is the `GetTargetedProject` call. In NET 8 we choose an arbitrary project in the solution as opposed to the executable project. This means the publish profile import could be non-deterministic unless further changes are made, and those changes, while not an extreme amount of work, also will need to be very well thought-through to prevent bugs

Reply:
interesting - so the API changed from "if I'm called with a solution I return null" to "if I'm called with a solution I pick a project" in 8? That's good to know - for publish specifically we'd want to not run at all if the argument is a solution, so I can change this call to explicitly check that condition before calling `GetTargetedProject`.

Reply:
Yeah. Docstring is updated here: https://github.com/dotnet/sdk/blob/0e9ba20184689a18972aa755edf522ee5c7fa9c8/src/Cli/dotnet/ReleasePropertyProjectLocator.cs#L120